### DISCUSSION TITLE
pdf.js in Angular - memory leak

### DISCUSSION THREAD

**Discussion by sergiubologa** [stephanrauh/ngx-extended-pdf-viewer#1544] [Q&A]
Hello! I've just found your library and looks very nice. I've integrated pdf.js in one of my Angular projects at work and I have a memory leak I can't get rid of so I thought I'd find the answer in your code base. 

I have an angular component that uses pdf.js. The user can navigate through a list of actions, selecting an action loads this component which displays the PDF. All nice and dandy, until we navigate faster through the documents. So it might happen that we call the destroy function while the document is rendered by pdf.js.

Now, the main problem is that the `PDFDocumentLoadingTask.destroy()` function in pdf.js is an async function:
```js
async destroy() {
    this.destroyed = true;
    await this._transport?.destroy();
    this._transport = null;

    if (this._worker) {
      this._worker.destroy();

      this._worker = null;
    }
  }
```

and I call it like this, from ngOnDestroy:
```js
async ngOnDestroy(): Promise<void> {
    await this._pdfDocumentLoadingTask.destroy();
}
```

The problem I identified is that sometimes `this._worker.destroy();` is never called or called very late (5-10 sec after `await this._transport?.destroy();` is called). At this time the new document will be already loaded since the Angular component doesn't wait for the destroy to finish. This happens when working with big documents in terms of resolution. With small documents, it works pretty well.

Also, looking at your codebase I saw that you wrapped the entire pdf.js code to run outside the Angular zone. Can you please tell me the reasoning for this?

Did you have a similar issue?

**Comment by stephanrauh** [comment]
People stopped complaining about memory leaks a long time ago, so I assume I've mostly solved the problem. It's not perfect, but the remaining memory leaks (if there are any) don't seem to be a problem. My `ngOnDestroy` methods is quite a few lines long. Feel free to examine the code and to reuse it:

https://github.com/stephanrauh/ngx-extended-pdf-viewer/blob/main/projects/ngx-extended-pdf-viewer/src/lib/ngx-extended-pdf-viewer.component.ts#L1624-L1684

I've tried to reproduce your delay, but I didn't manage. In any case, it's new to me. That doesn't mean it's not a problem in my library. I simply don't know because nobody's complained about it yet. It'd be nice to find out when and why `_worker.destroy()` is called late. 

I'd like to ask you to send me a reproducer - but since you're not using my library, that's demanding too much. So maybe I should create a reproducer myself, possibly using my favorite 75 MB poster (https://www.obwb.ca/library/okanagan-basin-waterscape-poster/), but I'm afraid I'm short of time during the next three weeks. Ping me in late October if you want me to investigate the topic thoroughly.

**Reply by sergiubologa** [reply]
Thank you for your reply. 

Can you please tell me the reasoning for wrapping the entire pdf.js code to run outside ngZone? I mean why not just the specific functions that call pdfjs? Was this part of a memory leak fix?

**Reply by sergiubologa** [reply]
I'll make a simple project to reproduce the issue and I'll send it over.

**Reply by stephanrauh** [reply]
As for running the code outside the Angular zone: that's nothing to do with memory. It's a performance optimization. pdf.js sends an incredible number of events, promises, and asynchronous code, and zone.js wraps almost everything. That results in a significant performance penalty. That's the first reason why I tried to talk zone.js into ignoring pdf.js. The other reason is stack traces. Even without zone.js, it's hard to read a pdf.js stacktrace because almost everything in pdf.js runs asynchronously. zone.js adds a dozen lines to every function call. Even worse, the developer tools of Chrome have the nice "Event Listeners" window - but it's useless if every event points to zone.js.

I didn't manage to run the entire code outside Angular. Zone.js still intercepts way to many events. It's still a nuisance when debugging. But at least, performance has improved by 30%, give or take a few, depending on your document.

**Reply by sergiubologa** [reply]
Hopefully, this won't be necessary anymore with Angular 15 if you opt out of ngZone.

Thanks so much for the explanations. I'll come back with a project to reproduce the memory leak.

**Reply by stephanrauh** [reply]
AFAIK, you can opt out of ngZone even today. It's not even difficult. It's just that the price tag is so heavy nobody does it. You've piqued my curiosity about Angular 15!

**Reply by sergiubologa** [reply]
Looks like I found the problem, it's a bug in Chromium (affects latest Edge and Chrome browsers). Details here: https://bugs.chromium.org/p/chromium/issues/detail?id=356205

**Reply by stephanrauh** [reply]
Awesome! BTW, I'm afraid you didn't get my last mail. For some reason unknown, it was sent back to me ("The email account that you tried to reach is over quota").

**Reply by sergiubologa** [reply]
Yes, sorry. My Google account went out of free usage storage and I didn't receive emails for a few days. Should be ok now.
